java/code/src/com/suse/oval/ovaldownloader/OVALDownloader.java

```diff
@@ -0,0 +1,178 @@
+
+    /**
+     * Default constructor
+     *
+     * @param configIn A configuration object that corresponds to {@code oval.config.json}
+     * */
+    public OVALDownloader(OVALConfig configIn) {
+        this.config = configIn;
+    }
+
+    /**
+     * Download the OVAL data that correspond to the given {@code osFamily} and {@code osVersion} and cache
+     * them for future access.
+     *
+     * @param osFamily the family of the Operating system (OS) to download OVAL data for.
+     * @param osVersion the version of the Operating system (OS) to download OVAL data for.
+     * @return {@link OVALDownloadResult}
+     * */
+    public OVALDownloadResult download(OsFamily osFamily, String osVersion) throws IOException {
+        Optional<OVALSourceInfo> sourceInfoOpt = config.lookupSourceInfo(osFamily, osVersion);
+        if (sourceInfoOpt.isEmpty()) {
+            throw new IllegalArgumentException(
+                    String.format(
+                            "OVAL sources for '%s' '%s' is not configured. Please ensure OVAL is configured " +
+                                    "correctly oval.config.json",
+                            osFamily, osVersion));
+        }
+
+        return doDownload(sourceInfoOpt.get());
+    }
+
+    /**
+     * A helper method for download OVAL data.
+     * */
+    private OVALDownloadResult doDownload(OVALSourceInfo sourceInfo)
+            throws IOException {
+        OVALDownloadResult result = new OVALDownloadResult();
+
+        String vulnerabilityInfoSource = sourceInfo.getVulnerabilitiesInfoSource();
+        String patchInfoSource = sourceInfo.getPatchInfoSource();
+
+        if (StringUtils.isNotBlank(vulnerabilityInfoSource)) {
+            File vulnerabilityFile = downloadOVALFile(vulnerabilityInfoSource);
+            result.setVulnerabilityFile(vulnerabilityFile);
+        }
+
+        if (StringUtils.isNotBlank(patchInfoSource)) {
+            File patchFile = downloadOVALFile(patchInfoSource);
+            result.setPatchFile(patchFile);
+        }
+
+        return result;
+    }
+
+    private File downloadOVALFile(String vulnerabilityInfoSource) throws IOException {
+        URL vulnerabilityInfoURL = new URL(vulnerabilityInfoSource);
+        String vulnerabilityInfoOVALFilename = FilenameUtils.getName(vulnerabilityInfoURL.getPath());
+        File vulnerabilityFile =
+                new File(DOWNLOAD_PATH + vulnerabilityInfoOVALFilename);
+        // Start downloading
+        FileUtils.copyURLToFile(vulnerabilityInfoURL, vulnerabilityFile, 10_000, 10_000);
+
+        return decompressIfNeeded(vulnerabilityFile);
+    }
+
+    private File decompressIfNeeded(File file) {
+        OVALCompressionMethod compressionMethod = getCompressionMethod(file.getName());
+        File uncompressedOVALFile;
+        if (compressionMethod == OVALCompressionMethod.BZIP2) {
+            uncompressedOVALFile = new File(file.getPath().replace(compressionMethod.extension(), ""));
```

Comment:
this replace is strange. Why not just remove the extension? Normally the files have the format `...oval.xml.bz2`. If we remove the `.bz2` we get the file named. With the replace, we could be changing the file name since the string can appear on any place.

Reply:
The goal was to remove the extension but as you said this would remove the extension string if it occurred inside the filename. It's fixed now.